sdk/program/src/stake/state.rs

```diff
@@ -564,9 +567,18 @@ impl Stake {
         Ok(new)
     }
 
-    pub fn deactivate(&mut self, epoch: Epoch) -> Result<(), StakeError> {
+    pub fn deactivate(
+        &mut self,
+        epoch: Epoch,
+        deactivation_flag: MustFullyActivateBeforeDeactivationIsPermittedFlag,
+    ) -> Result<(), StakeError> {
         if self.delegation.deactivation_epoch != std::u64::MAX {
             Err(StakeError::AlreadyDeactivated)
+        } else if deactivation_flag == 1 && epoch <= self.delegation.activation_epoch {
+            // when deactivation_flag is set, deactivation is only permitted
```

Comment:
oh so I remember the context behind this comment:
> The next problem will be that StakeInstruction::Deactivate currently doesn't receive the StakeHistory sysvar, and that sysvar is not gettable. So when MustFullyActivateBeforeDeactivationIsPermittedFlag is set, there's no way for Deactivate to determine if the stake account is now fully activated. Making StakeHistory gettable is probably preferable (from a non- client breaking pov) to adding StakeHistory as a required account to Deactivate when that flag is set.

Checking for `activation_epoch` + 1 is not sufficient in the case where more than 25% of the stake is moving at once. Doomsday: 100% of the stake is redelegated in the same epoch somehow. Even so, only 25% of the stake will actually change state in the next epoch. But the simple check here means that the remaining 75% could be deactivated once the current epoch is beyond `activation_epoch`

Reply:
Interesting. 
I wonder how does the current delegate code prevent 75% of the stake delegation from previous epoch to be deactivated?

Reply:
> StakeInstruction::Deactivate currently doesn't receive the StakeHistory sysvar, and that sysvar is not gettable

Oh wait. This comment is not true!

We can do something like this to test for `status.activating == 0` before permitting the deactivation instead of `epoch <= self.delegation.activation_epoch`:
https://github.com/solana-labs/solana/blob/9afe8ebbef83154fee9e343ec7c9663e84c2b9ce/programs/stake/src/stake_state.rs#L958-L965

Reply:
yeah. sound good.